frontend/src/pages/projects/screens/detail/notebooks/useNotebookImageData.ts

```diff
@@ -24,24 +24,83 @@ export const getNotebookImageData = (
     notebook.metadata.annotations?.['notebooks.opendatahub.io/last-image-selection']?.split(':') ??
     [];
 
-  // Fallback for non internal registry clusters
-  const imageStream =
-    images.find((image) => image.metadata.name === imageName) ||
-    images.find((image) =>
-      image.spec.tags
-        ? image.spec.tags.find(
-            (version) =>
-              version.from?.name === container.image &&
-              image.metadata.name === lastImageSelectionName,
-          )
-        : false,
-    );
-
-  // if the image stream is not found, consider it deleted
+  const notebookImageInternalRegistry = getNotebookImageInternalRegistry(
+    notebook,
+    images,
+    imageName,
+    versionName,
+  );
+  if (
+    notebookImageInternalRegistry &&
+    notebookImageInternalRegistry.imageAvailability !== NotebookImageAvailability.DELETED
+  ) {
+    return notebookImageInternalRegistry;
+  }
+  const notebookImageNoInternalRegistry = getNotebookImageNoInternalRegistry(
+    notebook,
+    images,
+    lastImageSelectionName,
+    versionName,
+    imageName,
+  );
+  if (
+    notebookImageNoInternalRegistry &&
+    notebookImageNoInternalRegistry.imageAvailability !== NotebookImageAvailability.DELETED
+  ) {
+    return notebookImageNoInternalRegistry;
+  }
+  const notebookImageNoInternalRegistryNoSHA = getNotebookImageNoInternalRegistryNoSHA(
+    notebook,
+    images,
+    lastImageSelectionTag,
+    versionName,
+    imageName,
+  );
+  if (
+    notebookImageNoInternalRegistryNoSHA &&
+    notebookImageNoInternalRegistryNoSHA.imageAvailability !== NotebookImageAvailability.DELETED
+  ) {
+    return notebookImageNoInternalRegistryNoSHA;
+  }
+  return {
+    imageAvailability: NotebookImageAvailability.DELETED,
+    imageDisplayName:
+      notebookImageInternalRegistry?.imageDisplayName ||
+      notebookImageNoInternalRegistry?.imageDisplayName ||
+      notebookImageNoInternalRegistryNoSHA?.imageDisplayName,
+  };
+};
+
+const useNotebookImageData = (notebook?: NotebookKind): NotebookImageData => {
+  const { dashboardNamespace } = useNamespaces();
+  const [images, loaded, loadError] = useImageStreams(dashboardNamespace, true);
+
+  return React.useMemo(() => {
+    if (!loaded || !notebook) {
+      return [null, false, loadError];
+    }
+    const data = getNotebookImageData(notebook, images);
+
+    if (data === null) {
+      return [null, false, loadError];
+    }
+
+    return [data, true, undefined];
+  }, [images, notebook, loaded, loadError]);
+};
+
+const getNotebookImageInternalRegistry = (
+  notebook: NotebookKind,
+  images: ImageStreamKind[],
+  imageName: string,
+  versionName: string,
+): NotebookImageData[0] => {
+  const imageStream = images.find((image) => image.metadata.name === imageName);
+
   if (!imageStream) {
     // Get the image display name from the notebook metadata if we can't find the image stream. (this is a fallback and could still be undefined)
```

Comment:
We are repeating this in three different parts, maybe we can convert this into a function? (the part inside of the conditional)

Reply:
yeah, I second that - there is just one line difference that can be driven by some function parameter